### PR TITLE
Configuration, MinGW/Cygwin and other cleanups + some bug fixes

### PR DESCRIPTION
**Pull request by ampli**
This started as a cleanup branch, but more problems have been fixed.
To save time I didn't separate the fixes to different PRs.

- Configuration cleanup (some are also related to MinGW/Cygwin):
-- Remove configuration references to `argp` (`link-generator` is not built if the `argp` package is missing). 
-- configure.ac: Replace `AM_PROG_LEX` with `AC_PROG_LEX` + check if `lex`/`flex` found (issue #1270).
-- configure.ac: Double quote `$with_regexlib`.
-- configure.ac: Check if `pkg-config` has been found (issue #1270).
-- configure.ac: Avoid using `sed -i` (non-portable, and existing code in `configure is buggy).
-- configure.ac: Notify a clear error on undefined `PKG_` macro.
-- configure.ac: Replace `echo`/`$ECHO` with `AS_ECHO`
-- configure.ac: Don't display backslashes in the Definitions printout (for readability).
-- configure.ac: Word around buggy `SWIG_LIB` (prevented python bindings build on MinGW).

- Cygwin/MinGW:
-- MinGW: Use %lu for DWORD (prevent warnings).
-- `is*()`: Use unsigned char (prevent a lot of warnings).
-- `valid_dialect_name()`: Use unsigned char (prevent a lot of warnings + possible crash on a bad dict).
-- Remove references to python2 (in documents/comments).
-- mingw/README-MinGW64.md: Fix typo `-ssp` -> `-lssp`.
-- strndupa(): Avoid a warning due to `-Wstringop-truncation` on `strncpy()`.
-- spellcheck-hun.c: Fix HUN_THREAD_PROTECT misdefinition (build failure).
-- tests.py: Fix `%p` matching in `test_low_level_exp` (test failure on Windows).
-- README-MinGW64.md: Add the package name which provides `flex`.
-- `lg_isatty()`: Clarify cygwin-style tty name (in a comment).
-- `lg_isatty()`: Cygwin tty: Don't check filename ending characters (bug fix for MinGW).
-- `lg_isatty()`: For "completeness", use the modern error value.
-- `fget_input_string()`: Fix condition for console reading (Cygwin/MinGW) (bug fix).
-- MinGW/Cygwin: Update the `link-parser` invocation scripts (make them much better).

- General fixes:
-- histogram.h: Remove redundant `typedef count_t` (PR #1267).
-- `valid_dialect_name()`: Use unsigned char (prevent a lot of warnings + possible crash on a bad dict).

- Notes:
-- After these fixes, everything runs fine for MinGW/Cygsys. (including all the tests).
-- For some (unknown to me) reason `link-generator`, when compiled with MSVC (PR #1269) or MinGW, runs very much slower than on Linux and Cygwin.

### CODE REVIEW

**Comment by ampli** [comment]
> Some checks were not successful

I will push a fix for that.

File: configure.ac [hunk 1]
```diff
@@ -843,10 +854,8 @@ AC_SUBST(pythondir)
 # of compiler warnings, so it got discarded here. Now we generate only
 # python3 bindings, but since we don't use this definition, it is still
 # discarded here.
-# Note: The v command allows to differentiate between the two flavors of sed.
-edit_in_place=-i
-sed v </dev/null 2>/dev/null || edit_in_place="-i ''"
-sed $edit_in_place '/HAVE_PYTHON/d' confdefs.h
+# Note: "sed -i" cannot be sued because it is not portable.
```
Comment (linas) [inline configure.ac:857]: `used` not `used`

File: configure.ac [hunk 2]
```diff
@@ -982,7 +988,7 @@ AC_FUNC_STRERROR_R
 
 dnl Save the compilation definitions for an extended version printout
 AC_OUTPUT_MAKE_DEFS()
-LG_DEFS=`$ECHO "$DEFS" | $SED 's/\\\\//g'`
+LG_DEFS=`AS_ECHO["$DEFS"] | $SED 's/\\\\//g'`
 AC_SUBST(LG_DEFS)
```
Comment (linas) [inline configure.ac:992]: what is the difference between `$SED` here and naked `sed` above?

**Review by linas — APPROVED** [review]
The `link-parser.bat` needs to be removed from `Makefile.am` too

**Pull request merged**